torch/csrc/jit/type.cpp

```diff
@@ -85,4 +85,48 @@ ListTypePtr ListType::ofFloats() {
   return value;
 }
 
+at::optional<TypePtr> unifyTypes(const TypePtr& t1, const TypePtr& t2) {
+  //cases that t1 == t2, or t1 is a type refinement of t2 and vice versa
+  if (t1->isSubtypeOf(t2)) {
+    return t2;
+  } else if (t2->isSubtypeOf(t1)) {
+    return t1;
+  }
+
+  // cases in which two types share a supertype
+  if (t1->isSubtypeOf(NumberType::get()) && t2->isSubtypeOf(NumberType::get())) {
+    return static_cast<TypePtr>(NumberType::get());
```

Comment:
Hmm I've been thinking about it and I'm not 100% sure if we really want to allow returning `NumberType`s from loops and conditionals. E.g. we don't have an op for `aten::add(Int, Number)`.

Reply:
I think not allowing NumberTypes until we have the proper op support for Numbers is reasonable